ref_arch/multi_tenant.rst

```diff
@@ -0,0 +1,511 @@
+
+  for dataset in companies campaigns ads clicks impressions; do
+    curl -O https://examples.citusdata.com/mt_ref_arch/${dataset}.csv
+  done
+
+.. note::
+
+  **If you are using Docker,** you should use the :code:`docker cp` command to copy the files into the Docker container.
+
+  .. code-block:: bash
+
+    for dataset in companies campaigns ads clicks impressions; do
+      docker cp ${dataset}.csv citus_master:.
+    done
+
+Being an extension of PostgreSQL, Citus supports bulk loading with the COPY command. Use it to ingest the data you downloaded, and make sure that you specify the correct file path if you downloaded the file to some other location.
+
+.. code-block:: psql
+
+  \copy companies from 'companies.csv' with csv;
+  \copy campaigns from 'campaigns.csv' with csv;
+  \copy ads from 'ads.csv' with csv;
+  \copy clicks from 'clicks.csv' with csv;
+  \copy impressions from 'impressions.csv' with csv;
+
+Integrating Applications
+------------------------
+
+Here's the good news: once you have made the slight schema modification outlined earlier, your application can scale with very little work. You'll just connect the app to Citus and let the database take care of keeping the queries fast and the data safe.
+
+Any application queries or update statements which include a filter on :code:`company_id` will continue to work exactly as they are. As mentioned earlier, this kind of filter is common in multi-tenant apps. When using an Object-Relational Mapper (ORM) you can recognize these queries by methods such as :code:`where` or :code:`filter`.
+
+ActiveRecord:
+
+.. code-block:: ruby
+
+  Impression.where(company_id: 5).count
+
+Django:
+
+.. code-block:: py
+
+  Impression.objects.filter(company_id=5).count()
+
+Basically when the resulting SQL executed in the database contains a :code:`WHERE company_id = :value` clause on every table (including tables in JOIN queries), then the query will be routed to a single node and will work fine.
+
+Rather than track down each query in your application to make sure it filters properly, you can use our library for Rails (one for Django is in progress as well). It allows you to surround queries in a tenant scope which automatically filters them, even complicated ones. Check out our :ref:`rails_migration` section for details.
+
+This guide is framework-agnostic, so we'll point out some Citus features using SQL. Use your imagination for how these statements would be expressed in your language of choice.
+
+When a Citus sends a query to a single node for execution (called "pushing down" the query), then all SQL functionality is available. The node is an ordinary PostgreSQL server after all. When people scale applications with NoSQL databases they suffer the lack of transactions and joins. However per-tenant transactions work fine in Citus:
+
+::
+
+  -- transactionally remove campaign 46 and all its ads etc
+
+  BEGIN;
+  DELETE from campaigns where id = 46 AND company_id = 5;
```

Comment:
Actually might be cooler to use `ON DELETE CASCADE` in our foreign key, and pick another set of statements for this transaction.